projects/PdfDing/services/pdfding/module.nix

```diff
@@ -0,0 +1,464 @@
+          ${loadCreds}
+          exec ${cfg.package}/bin/pdfding-start ${builtins.toString cfg.gunicorn.extraArgs}
+        '';
+        EnvironmentFile = cfg.envFiles;
+        NoNewPrivileges = true;
+        PrivateTmp = true;
+        PrivateDevices = true;
+        ProtectSystem = "strict";
+        ProtectHome = true;
+        ReadWritePaths = [ cfg.dataDir ];
+        Restart = "on-failure";
+        RestartSec = "5s";
+      };
+    };
+
+    systemd.services.pdfding-background = lib.mkIf (cfg.consume.enable || cfg.backup.enable) {
+      description = "PdfDing Background Tasks (Huey)";
+      after = [ "pdfding.service" ];
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        Type = "exec";
+        User = cfg.user;
+        Group = cfg.group;
+        WorkingDirectory = cfg.dataDir;
+        EnvironmentFile = cfg.envFiles;
+        ExecStart = pkgs.writeShellScript "exec-start" ''
+          ${loadCreds}
+          exec ${cfg.package}/bin/pdfding-manage run_huey;
+        '';
+
+        NoNewPrivileges = true;
+        PrivateTmp = true;
+        PrivateDevices = true;
+        ProtectSystem = "strict";
+        ProtectHome = true;
+        ReadWritePaths = [ cfg.dataDir ];
+        Restart = "on-failure";
+        RestartSec = "5s";
+        TimeoutStopSec = 30;
+      };
+    };
+
+    # postgres setup
+    services.postgresql.enable = lib.mkDefault (usePostgres && cfg.database.createLocally);
+
+    # copied from keycloak module in nixpkgs
+    systemd.services.pdfdingPostgreSQLInit = lib.mkIf (usePostgres && cfg.database.createLocally) {
+      after = [ "postgresql.target" ];
+      before = [ "pdfding.service" ];
+      bindsTo = [ "postgresql.target" ];
+      path = [ config.services.postgresql.package ];
+      serviceConfig = {
+        Type = "oneshot";
+        RemainAfterExit = true;
+        User = "postgres";
+        Group = "postgres";
+        LoadCredential = [ "db_password:${cfg.database.passwordFile}" ];
+      };
+      script =
+        # bash
+        ''
+          set -o errexit -o pipefail -o nounset -o errtrace
+          shopt -s inherit_errexit
+
+          create_role="$(mktemp)"
+          trap 'rm -f "$create_role"' EXIT
+
+          # escape any single quotes by adding additional single
+          # quotes after them, following the rules laid out here:
+          # https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-CONSTANTS
+          POSTGRES_PASSWORD="$(<"$CREDENTIALS_DIRECTORY/db_password")"
+          POSTGRES_PASSWORD="''${POSTGRES_PASSWORD//\'/\'\'}"
+
+          echo "CREATE ROLE pdfding WITH LOGIN PASSWORD '$POSTGRES_PASSWORD' CREATEDB" > "$create_role"
+          psql -tAc "SELECT 1 FROM pg_roles WHERE rolname='pdfding'" | grep -q 1 || psql -tA --file="$create_role"
+          psql -tAc "SELECT 1 FROM pg_database WHERE datname = 'pdfding'" | grep -q 1 || psql -tAc 'CREATE DATABASE "pdfding" OWNER "pdfding"'
+        '';
```

Comment:
I would try to avoid this kind of hacking and use what postgresql options are currently providing even in cost of loosing some configuration options. This kind of scripts can be hard to maintain and we can end up having many of them in multiple projects.

Also, I would suggest to use unix socket for local db connection.

Reply:
Yes, there is a stalled pr for this exact hack, https://github.com/NixOS/nixpkgs/pull/326306

My understanding is, I will keep maintaining the upstream nixos module (once upstreamed to nixpkgs) and I don't feel like this particular hack is any harder to maintain than the rest of pdfding.

As for unix socket connection, I will add it later if its not a blocker.

Reply:
Pdfding defaults to some password if not provided, https://github.com/search?q=repo%3Amrmn2%2FPdfDing+POSTGRES_PASSWORD&type=code

Django documentation says to use `HOST` attribute https://docs.djangoproject.com/en/6.0/ref/settings/#host to be set empty for nudging it to use a unix domain socket for postgresql. I will check if that works.

Reply:
Done.

Tests pass

```
$ git log -1
commit 21eacafe7439e859183636247ad01ff243e08c6b
$ nom build -L .#checks.x86_64-linux.projects/PdfDing/nixos/tests/{basic,postgres,minio,e2e}{.driver,.driverInteractive,}

$ # rm result*
```